node/_fs/_fs_opendir.ts

```diff
@@ -0,0 +1,91 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+import Dir from "./_fs_dir.ts";
+import { Buffer } from "../buffer.ts";
+import { getValidatedPath } from "../internal/fs/utils.mjs";
+import { denoErrorToNodeError } from "../internal/errors.ts";
+import { validateFunction } from "../internal/validators.mjs";
+import { promisify } from "../internal/util.mjs";
+
+type Options = {
+  encoding?: string;
+  bufferSize?: number;
+};
+type Callback = (err?: Error | null, dir?: Dir) => void;
+
+function _validateFunction(callback: unknown): asserts callback is Callback {
+  validateFunction(callback, "callback");
+}
+
+function checkBufferSize(value: number, name: string) {
+  if (!Number.isInteger(value) || value < 1 || value > 4294967295) {
+    throw new RangeError(
+      `The value of "${name}" is out of range. It must be >= 1 && <= 4294967295. Received ${value}`,
+    );
+  }
+}
+
+export function opendir(
+  path: string | Buffer | URL,
+  options: Options | Callback,
+  callback?: Callback,
+) {
+  callback = typeof options === "function" ? options : callback;
+  _validateFunction(callback);
+
+  path = getValidatedPath(path).toString();
+
+  options = typeof options === "object" ? options : {};
+  options = Object.assign({
+    encoding: "utf8",
+    bufferSize: 32,
+  }, options);
+
+  try {
+    /** Throws if encoding is invalid */
+    new TextDecoder(options.encoding);
```

Comment:
I think this needs to be `validateEncoding()` call, which is defined in `node/internal/validators.mjs`. You can pass, for example, `encoding: "binary"` in opendir in node.js, but `new TextDecoder("binary")` throws.

Reply:
Ah, ok. Am I supposed to use `path` for the `data` parameter?

Reply:
Oops, sorry `validateEncoding` doesn't look correct for this according to the thrown error message. Probably `assertEncoding` (in `node/internal/fs/utils.mjs`) is the right choice here.